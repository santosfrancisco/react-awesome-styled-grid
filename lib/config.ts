import { Config, ConfigProps, GridBreakpoints, Media } from "@site/types";
import { css, CSSObject } from "styled-components";

export const CUSTOM_CONF = "awesomegrid";
export const DIMENSIONS: GridBreakpoints[] = ["xs", "sm", "md", "lg", "xl"];
export const BASE_CONF: Config = {
  media: null,
  mediaQuery: "only screen",
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: "full", // 'full' = max-width: 100%
    sm: "full", // 'full' = max-width: 100%
    md: "full", // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
};

const configs = [];
const hasCustomConf = (props: ConfigProps) => {
  if (!props?.theme) return {};
  return JSON.stringify(props.theme[CUSTOM_CONF] || {});
};
const resolveConfig = (props: ConfigProps) => {
  let themeConf = {};
  if (props?.theme) {
    themeConf = props.theme[CUSTOM_CONF] || {};
  }

  const conf: Config = {
    ...BASE_CONF,
    ...themeConf
  };

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint];
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(" and ")
    );
    return media;
  }, {} as Media);

  return conf;
};

export default function config(props?: ConfigProps): Config {
  const customConf = hasCustomConf(props);
  if (configs[0] === customConf) {
    return configs[1];
  }

  const conf = resolveConfig(props);

  configs[0] = customConf;
  configs[1] = conf;

  return conf;
}

function makeMedia(media: string) {
  return (...args: CSSObject[] | TemplateStringsArray) => css`
    @media ${media} {
      ${css.apply(null, [...args])}
    }
  `;
}
