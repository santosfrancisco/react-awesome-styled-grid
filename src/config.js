import { css } from 'styled-components'

const CUSTOM_CONF = 'awesomegrid'
export const DIMENSIONS = ['xs', 'sm', 'md', 'lg', 'xl']
export const BASE_CONF = {
  mediaQuery: 'only screen',
  columns: {
    xs: 4,
    sm: 8,
    md: 8,
    lg: 12,
    xl: 12
  },
  gutterWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  paddingWidth: {
    xs: 1,
    sm: 1,
    md: 1.5,
    lg: 1.5,
    xl: 1.5
  },
  container: {
    xs: 'full', // 'full' = max-width: 100%
    sm: 'full', // 'full' = max-width: 100%
    md: 'full', // 'full' = max-width: 100%
    lg: 90, // max-width: 1440px
    xl: 90 // max-width: 1440px
  },
  breakpoints: {
    xs: 1,
    sm: 48, // 768px
    md: 64, // 1024px
    lg: 90, // 1440px
    xl: 120 // 1920px
  }
}

const configs = []
const hasCustomConf = props => JSON.stringify((props.theme && props.theme[CUSTOM_CONF]) || {})
const resolveConfig = props => {
  const themeConf = (props.theme && props.theme[CUSTOM_CONF]) || {}

  const conf = {
    ...BASE_CONF,
    ...themeConf
  }

  conf.media = Object.keys(conf.breakpoints).reduce((media, breakpoint) => {
    const breakpointWidth = conf.breakpoints[breakpoint]
    media[breakpoint] = makeMedia(
      [
        conf.mediaQuery,
        breakpointWidth >= 0 && `(min-width: ${breakpointWidth}rem)`
      ]
        .filter(Boolean)
        .join(' and ')
    )
    return media
  }, {})

  return conf
}

export default function config (props = {}) {
  const customConf = hasCustomConf(props)
  if (configs[0] === customConf) {
    return configs[1]
  }

  const conf = resolveConfig(props)

  configs[0] = customConf
  configs[1] = conf

  return conf
}

function makeMedia (media) {
  return (...args) => css`
    @media ${media} {
      ${css(...args)}
    }
  `
}
