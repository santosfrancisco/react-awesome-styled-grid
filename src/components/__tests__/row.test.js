import React from 'react'
import renderer from 'react-test-renderer'
import { DIMENSIONS, BASE_CONF } from '../../config'
import Row from '../grid/row'
import 'jest-styled-components'

describe('<Row />', () => {
  it('should render default style correctly', () => {
    const tree = renderer.create(<Row />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should have a reverse direction for each media', () => {
    const tree = renderer.create(<Row reverse />).toJSON()

    expect(tree).toHaveStyleRule(
      'flex-direction', `row-reverse`)
  })

  it('should have a reverse direction for each media', () => {
    const tree = renderer.create(<Row reverse />).toJSON()

    expect(tree).toHaveStyleRule(
      'flex-direction', `row-reverse`)
  })

  it('should have a reverse direction for specific screen', () => {
    const tree = renderer.create(<Row reverse={['md', 'lg']} />).toJSON()
    const { breakpoints } = BASE_CONF

    expect(tree).toHaveStyleRule(
      'flex-direction', `row-reverse`, {
        media: `only screen and (min-width: ${breakpoints['md']}rem)`
      })

    expect(tree).toHaveStyleRule(
      'flex-direction', `row-reverse`, {
        media: `only screen and (min-width: ${breakpoints['lg']}rem)`
      })
  })

  it('should have different style when it`s debug props is true', () => {
    const tree = renderer.create(<Row debug />).toJSON()
    expect(tree).toHaveStyleRule('background-color', '#5901ad40')
    expect(tree).toHaveStyleRule('outline', '#fff solid 1px')
  })

  it('should pass every props', () => {
    const tree = renderer.create(<Row id='uniqueID' lol='anotherprop' />)
    const testInstance = tree.root
    expect(testInstance.findByType(Row).props.id).toBe('uniqueID')
    expect(testInstance.findByType(Row).props.lol).toBe('anotherprop')
  })

  it('should have the css rule justify-content center', () => {
    const props = {
      justify: 'center'
    }
    const tree = renderer.create(<Row {...props} />).toJSON()

    expect(tree).toHaveStyleRule('justify-content', 'center')
  })

  it('should have the css rule justify-content center only in MD screen', () => {
    const props = {
      justify: {
        md: 'center'
      }
    }

    const tree = renderer.create(<Row {...props} />).toJSON()

    const { breakpoints } = BASE_CONF

    DIMENSIONS.forEach(d => {
      expect(tree).toHaveStyleRule(
        'justify-content', props.justify[d], {
          media: `only screen and (min-width: ${breakpoints[d]}rem)`
        }
      )
    })
  })

  it('should have the css rule align-items center', () => {
    const props = {
      align: 'center'
    }
    const tree = renderer.create(<Row {...props} />).toJSON()

    expect(tree).toHaveStyleRule('align-items', props.align)
  })

  it('should have the css rule align-items center only in MD screen', () => {
    const props = {
      align: {
        md: 'center'
      }
    }

    const tree = renderer.create(<Row {...props} />).toJSON()

    const { breakpoints } = BASE_CONF

    DIMENSIONS.forEach(d => {
      expect(tree).toHaveStyleRule(
        'align-items', props.align[d], {
          media: `only screen and (min-width: ${breakpoints[d]}rem)`
        }
      )
    })
  })
})
