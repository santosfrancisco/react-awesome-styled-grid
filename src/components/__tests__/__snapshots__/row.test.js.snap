// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> should render default style correctly 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-width: 100%;
}

@media only screen and (min-width:1rem) {
  .c0 {
    margin-left: -0.5rem;
    margin-right: -0.5rem;
  }
}

@media only screen and (min-width:48rem) {
  .c0 {
    margin-left: -0.5rem;
    margin-right: -0.5rem;
  }
}

@media only screen and (min-width:64rem) {
  .c0 {
    margin-left: -0.75rem;
    margin-right: -0.75rem;
  }
}

@media only screen and (min-width:90rem) {
  .c0 {
    margin-left: -0.75rem;
    margin-right: -0.75rem;
  }
}

@media only screen and (min-width:120rem) {
  .c0 {
    margin-left: -0.75rem;
    margin-right: -0.75rem;
  }
}

<div
  className="c0"
/>
`;
